kernel/gic/src/mod.rs

```diff
@@ -0,0 +1,310 @@
+use core::convert::AsMut;
+
+use memory::{PageTable, BorrowedMappedPages, Mutable,
+PhysicalAddress, PteFlags, allocate_pages, allocate_frames_at};
+
+use static_assertions::const_assert_eq;
+use bitflags::bitflags;
+
+/// Physical addresses of the CPU and Distributor
+/// interfaces as exposed by the qemu "virt" VM.
+pub mod qemu_virt_addrs {
+    use super::*;
+
+    pub const GICD: PhysicalAddress = PhysicalAddress::new_canonical(0x08000000);
+    pub const GICC: PhysicalAddress = PhysicalAddress::new_canonical(0x08010000);
+    pub const GICR: PhysicalAddress = PhysicalAddress::new_canonical(0x080A0000);
+}
+
+/// Boolean
+pub type Enabled = bool;
+
+/// 24-bit unsigned integer
+///
+/// An u32 is used because there is no u24.
+pub type IntNumber = u32;
+
+/// 8-bit unsigned integer
+pub type Priority = u8;
+
+bitflags! {
+    /// The legacy (GICv2) way of specifying
+    /// multiple target CPUs
+    pub struct TargetList: u8 {
+        const CPU_0 = 1 << 0;
+        const CPU_1 = 1 << 1;
+        const CPU_2 = 1 << 2;
+        const CPU_3 = 1 << 3;
+        const CPU_4 = 1 << 4;
+        const CPU_5 = 1 << 5;
+        const CPU_6 = 1 << 6;
+        const CPU_7 = 1 << 7;
+        const ALL_CPUS = u8::MAX;
+    }
+}
+
+pub enum TargetCpu {
+    /// That interrupt must be handled by
+    /// a specific PE in the system
+    Specific(u32),
+    /// That interrupt can be handled by
+    /// any PE that is not busy with another,
+    /// more important task
+    AnyCpuAvailable,
+    GICv2TargetList(TargetList),
+}
+
+// = 4
+const U32BYTES: usize = core::mem::size_of::<u32>();
+// = 32
+const U32BITS: usize = U32BYTES * 8;
+
+#[repr(C)]
+#[derive(zerocopy::FromBytes)]
+pub struct GicMappedPage {
```

Comment:
I disagree with using `MemoryMappedGicRegisters` as it doesn't mention that the mapping is one page long, which to me is important as some interfaces (dist, redist for example) are more than one page long; they use multiple instances of this struct.

Reply:
The issue with `GicMappedPage` is two-fold:
1. it implies that it is the singular mapped page for the GIC, which may not be true (since there are multiple sets of single-page MappedPages objects for a GIC
2. It's not a MappedPage, it's the type that you are overlaying atop a MappedPages, i.e., what you are reinterpreting that memory as. Calling it a MappedPage in this context is actually incorrect, not just my preference.

Why does it matter how many pages it is? Take a look at the `ApicRegisters` type, which is 100% analogous to this type. If you really want to include that unnecessary detail in the type name, you could go with `GicRegistersPage` or something similar.

Reply:
Ok, I renamed it GicRegisters